enterprise/wsproxy/wsproxy_test.go

```diff
@@ -430,6 +434,173 @@ resourceLoop:
+	deploymentValues := coderdtest.DeploymentValues(t)
+	deploymentValues.Experiments = []string{
+		"*",
+	}
+
+	client, closer, api, _ := coderdenttest.NewWithAPI(t, &coderdenttest.Options{
+		Options: &coderdtest.Options{
+			DeploymentValues:         deploymentValues,
+			AppHostname:              "*.primary.test.coder.com",
+			IncludeProvisionerDaemon: true,
+			RealIPConfig: &httpmw.RealIPConfig{
+				TrustedOrigins: []*net.IPNet{{
+					IP:   net.ParseIP("127.0.0.1"),
+					Mask: net.CIDRMask(8, 32),
+				}},
+				TrustedHeaders: []string{
+					"CF-Connecting-IP",
+				},
+			},
+		},
+		LicenseOptions: &coderdenttest.LicenseOptions{
+			Features: license.Features{
+				codersdk.FeatureWorkspaceProxy: 1,
+			},
+		},
+	})
+	t.Cleanup(func() {
+		_ = closer.Close()
+	})
+
+	proxyURL, err := url.Parse("https://proxy.test.coder.com")
+	require.NoError(t, err)
+
+	// Create 6 proxy replicas.
+	const count = 6
+	var (
+		sessionToken = ""
+		proxies      = [count]coderdenttest.WorkspaceProxy{}
+		derpURLs     = [count]string{}
+	)
+	for i := range proxies {
+		proxies[i] = coderdenttest.NewWorkspaceProxy(t, api, client, &coderdenttest.ProxyOptions{
+			Name:     "best-proxy",
+			Token:    sessionToken,
+			ProxyURL: proxyURL,
+		})
+		if i == 0 {
+			sessionToken = proxies[i].Options.ProxySessionToken
+		}
+
+		derpURL := *proxies[i].ServerURL
+		derpURL.Path = "/derp"
+		derpURLs[i] = derpURL.String()
+	}
+
+	// Force all proxies to re-register immediately. This ensures the DERP mesh
+	// is up-to-date. In production this will happen automatically after about
+	// 15 seconds.
+	for i, proxy := range proxies {
+		err := proxy.RegisterNow(testutil.Context(t, testutil.WaitLong))
+		require.NoErrorf(t, err, "failed to force proxy %d to re-register", i)
+	}
+
+	createClient := func(t *testing.T, name string, derpURL string) (*derphttp.Client, <-chan derp.ReceivedPacket) {
+		t.Helper()
+
+		client, err := derphttp.NewClient(key.NewNode(), derpURLs[0], func(format string, args ...any) {
+			t.Logf(name+": "+format, args...)
+		})
+		require.NoError(t, err, "create client")
+		t.Cleanup(func() {
+			_ = client.Close()
+		})
+		err = client.Connect(testutil.Context(t, testutil.WaitLong))
+		require.NoError(t, err, "connect to DERP server")
+
+		ch := make(chan derp.ReceivedPacket, 64)
+		go func() {
+			defer close(ch)
+			for {
+				msg, err := client.Recv()
+				if err != nil {
+					t.Logf("Recv error: %v", err)
+					return
+				}
+				switch msg := msg.(type) {
+				case derp.ReceivedPacket:
+					ch <- msg
+					return
+				default:
+					// We don't care about other messages.
+				}
+			}
+		}()
+
+		return client, ch
+	}
+
+	sendTest := func(t *testing.T, dstKey key.NodePublic, dstCh <-chan derp.ReceivedPacket, src *derphttp.Client) {
+		t.Helper()
+
+		const msgStrPrefix = "test_packet_"
+		msgStr, err := cryptorand.String(64 - len(msgStrPrefix))
+		require.NoError(t, err, "generate random msg string")
+		msg := []byte(msgStrPrefix + msgStr)
```

Comment:
the prefix business seems superfluous if we're just going to assert that the packet data matches the msg.